sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,455 @@
+    }
+
+    private async Task LogRequestContent(PipelineRequest request, string? requestId, bool async, CancellationToken cancellationToken)
+    {
+        // If logging content is disabled, or the request content is null, or informational logs are not enabled, return
+        // Checking the log level prevents expensive operations from being executed
+        if (!_logContent || request.Content == null || !Singleton.IsEnabled(EventLevel.Informational, EventKeywords.All))
+        {
+            return; // nothing to log
+        }
+
+        // Convert binary content to bytes
+        using var memoryStream = new MaxLengthStream(_maxLength);
+        if (async)
+        {
+            await request.Content.WriteToAsync(memoryStream, cancellationToken).ConfigureAwait(false);
+        }
+        else
+        {
+            request.Content.WriteTo(memoryStream, cancellationToken);
+        }
+        var bytes = memoryStream.ToArray();
+
+        // Try to extract a text encoding from the headers
+        Encoding? requestTextEncoding = null;
+
+        if (request.Headers.TryGetValue("Content-Type", out var contentType) && contentType != null)
+        {
+            ContentTypeUtilities.TryGetTextEncoding(contentType, out requestTextEncoding);
+        }
+
+        // Log to event source
+        Singleton.RequestContent(requestId ?? string.Empty, bytes, requestTextEncoding);
+    }
+
+    private void LogResponse(PipelineResponse response, string? responseId, double elapsed)
+    {
+        bool isError = response.IsError;
+        if (isError)
+        {
+            Singleton.ErrorResponse(response, responseId ?? string.Empty, Sanitizer!, elapsed);
+        }
+        else
+        {
+            Singleton.Response(response, responseId ?? string.Empty, Sanitizer!, elapsed);
+        }
+    }
+
+    private async Task LogResponseContent(PipelineResponse response, string? responseId, bool bufferResponse, bool async, CancellationToken cancellationToken)
+    {
+        // If logging content is disabled, or the request content is null, or informational logs are not enabled, return.
+        // Checking the log level prevents expensive operations from being executed
+        // TODO - consider moving this down to LoggingStream
+        var logErrorResponseContent = response.IsError && Singleton.IsEnabled(EventLevel.Warning, EventKeywords.All);
+        var logResponseContent = Singleton.IsEnabled(EventLevel.Informational, EventKeywords.All) || logErrorResponseContent;
+
+        if (!_logContent || !logResponseContent)
+        {
+            return; // nothing to log
+        }
+
+        // Determine if the content stream should be wrapped in a logging stream, which logs the stream in blocks as it is being read
+        var wrapContent =
+            !bufferResponse // Content is not buffered - content should be in the content stream
+            && response.ContentStream != null // Content stream is available
```

Comment:
Let's add this to line 202 above -- if there's no content stream, there's no response content to log and we can exit the method early.  This will simplify logic downstream and make the code more clear and readable/maintainable.

Reply:
Yes, I wasn't sure if there would ever be a case where there is something in response.Content but not in response.ContentStream, but it sounds like no